src/system.rs

```diff
@@ -2,9 +2,105 @@
 
 use std::str::FromStr;
 use std::path::Path;
+use std::collections::HashMap;
 
 use utils::read_file;
 
+/// Gets a value from a hashmap and dereferences it.
+/// If the key doesn't exist, return `None`
+macro_rules! get_deref(
+    ($map:expr,$key:expr) => (
+        match $map.get($key) {
+            Some(e) => *e,
+            None => return None
+        }
+    )
+);
+
+#[derive(Debug)]
+pub struct VirtualMemory {
+    /// Amount of total memory
+    pub total: u64,
+
+    /// Amount of memory available for new processes
+    pub available: u64,
+
+    /// Percent of memory used
+    pub percent: f32,
+
+    /// Memory currently in use
+    pub used: u64,
+
+    /// Memory not being used
+    pub free: u64,
+
+    /// Memory currently in use
+    pub active: u64,
+
+    /// Memory that is not in use
+    pub inactive: u64,
+
+    /// Temporary storage for raw disk blocks
+    pub buffers: u64,
+
+    /// Memory used by the page cache
+    pub cached: u64,
+
+    /// Amount of memory consumed by tmpfs filesystems
+    pub shared: u64,
+}
+
+impl Default for VirtualMemory {
+    fn default() -> VirtualMemory {
```

Comment:
https://doc.rust-lang.org/nomicon/constructors.html says "While Rust provides a Default trait for specifying the moral equivalent of a default constructor, it's incredibly rare for this trait to be used." - so I'm not sure this is idiomatic.

Instead, I think there should be a `pub fn new` as an idiomatic constructor - probably taking total, available, shared, free, buffers, cached active and inactive as arguments, and calculating used and percent. Likewise for swap_memory.